client/src/util/typedCrossfilter/crossfilter.ts

```diff
@@ -8,32 +31,64 @@ import {
   lowerBoundIndirect,
   upperBoundIndirect,
 } from "./sort";
-import { makeSortIndex } from "./util";
-import { isAnyArray } from "../../common/types/arraytypes";
+import { makeSortIndex, NotImplementedError } from "./util";
 
-class NotImplementedError extends Error {
-  constructor(msg: string) {
-    super(msg);
+/**
+ * The result of a dimension selection
+ * @internal
+ */
+type DimensionSelectionResult = {
+  ranges: IntervalArray;
+  index: IndexArray | null;
+};
 
-    // Maintains proper stack trace for where our error was thrown (only available on V8)
-    if (Error.captureStackTrace) {
-      Error.captureStackTrace(this, NotImplementedError);
-    }
-  }
+/**
+ * Table of dimensions added to the crossfilter
+ * @internal
+ */
+interface Dimension {
+  id: number | undefined; // ID is often undefined when selection cache is clear
```

Comment:
```suggestion
  id?: number; // ID is often undefined when selection cache is clear
```

Reply:
No!  This is _not_ an optional field.  I want to enforce that it is always defined, even if the value is undefined.

I think you are confusing the (real) value `undefined` with the default behavior of object property access when a field is missing (eg, `foo.bar` evaluates to `undefined` when `bar` is missing).

Reply:
Sync'd with Bruce on this! The big upside is explicitness, where we require callsite to pass `undefined` for certain `key` instead of leaving them out as an optional argument:

E.g.,

```ts
function foo({ foo }: { foo: number | undefined }) { ... }

foo({}) // error
foo({ foo: undefined }) // good

function foo({ foo }: { foo?: number }) { ... }

foo({}) // good
foo({ foo: undefined }) // good
```